webpack.config.js

```diff
@@ -135,16 +135,19 @@ module.exports = {
 	entry: createEntrypoints(),
 	output: {
 		devtoolNamespace: 'wp',
-		filename: ( data ) => {
-			const { chunk } = data;
+		filename: ( pathData ) => {
+			const { chunk } = pathData;
 			const { entryModule } = chunk;
-			const { rawRequest } = entryModule;
+			const { rawRequest, rootModule } = entryModule;
 
-			/*
-			 * If the file being built is a Core Block's frontend file,
-			 * we build it in the block's directory.
-			 */
-			if ( rawRequest && rawRequest.includes( '/frontend.js' ) ) {
+			// When processing ESM files, the requested path
+			// is defined in `entryModule.rootModule.rawRequest`, instead of
+			// being present in `entryModule.rawRequest`.
+			// In the context of frontend files, they would be processed
+			// as ESM if they use `import` or `export` within it.
+			const request = rootModule?.rawRequest || rawRequest;
```

Comment:
I see it bring back the handling for `rawRequest` removed in #30341.

Reply:
It does. It interesting because I removed it as it seemed to be a legacy way of handling modules of a different kind, shown by the fact that checksumming builds made with or without the change rendered the same result — every source file was being built into the same path structure regardless of its type. 

The check becomes necessary again when building modules into arbitrary directories by using `rawRequest` as a starting point.